global var init from builtin

The following program `sample.go` triggers a panic:

```go
package main

import (
    "fmt"
)   

type T struct {
    Name string
}   

//var m = map[string]*T{} // ok
var m = make(map[string]*T) // fails

func main() {
    fmt.Println(m)
}
```

Expected result:
```console
$ go run ./sample.go
map[]
```

Got:
```console
$ yaegi ./sample.go
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xb0 pc=0x61a7ef]

goroutine 1 [running]:
github.com/containous/yaegi/interp.nodeType(0xc0001f0000, 0xc00014a3c0, 0xc000151200, 0x0, 0x0, 0xc00013fec0)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:329 +0x1ecf
github.com/containous/yaegi/interp.nodeType(0xc0001f0000, 0xc00014a3c0, 0xc000151100, 0x1, 0xc0001fc300, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:259 +0x1342
github.com/containous/yaegi/interp.(*Interpreter).gta.func1(0xc000150f00, 0xc0001bbd00)
        /home/marc/go/src/github.com/containous/yaegi/interp/gta.go:49 +0x5f6
github.com/containous/yaegi/interp.(*node).Walk(0xc000150f00, 0xc0001bbd58, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:111 +0xc9
github.com/containous/yaegi/interp.(*node).Walk(0xc000150e00, 0xc0001bbd58, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:115 +0x70
github.com/containous/yaegi/interp.(*node).Walk(0xc000150100, 0xc0001bbd58, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:115 +0x70
github.com/containous/yaegi/interp.(*Interpreter).gta(0xc0001f0000, 0xc000150100, 0xc000148098, 0x4, 0x1d, 0xc000148098, 0x4, 0xc000150100, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/gta.go:18 +0x127
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0001f0000, 0xc0001f6000, 0xa7, 0x2a7, 0xc0001f6000, 0xa7, 0x0, 0xc00013fb00)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:253 +0x104
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:77 +0x2f9
```

This issue is a blocker for #260

--- a/_test/map18.go
+++ b/_test/map18.go
@@ -1,0 +1,14 @@
+package main
+
+import "fmt"
+
+var foo = make([]int, 1)
+
+func main() {
+	for _, v := range foo {
+		fmt.Println(v)
+	}
+}
+
+// Output:
+// 0

--- a/_test/var7.go
+++ b/_test/var7.go
@@ -1,0 +1,18 @@
+package main
+
+import (
+	"fmt"
+)
+
+type T struct {
+	Name string
+}
+
+var m = make(map[string]*T)
+
+func main() {
+	fmt.Println(m)
+}
+
+// Output:
+// map[]

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -640,73 +640,17 @@ func (interp *Interpreter) cfg(root *node) ([]*node, error) {
 		case callExpr:
 			wireChild(n)
 			switch {
-			case isBuiltinCall(n):
+			case interp.isBuiltinCall(n):
 				n.gen = n.child[0].sym.builtin
 				n.child[0].typ = &itype{cat: builtinT}
-				switch n.child[0].ident {
-				case "append":
-					c1, c2 := n.child[1], n.child[2]
-					if n.typ = sc.getType(c1.ident); n.typ == nil {
-						if n.typ, err = nodeType(interp, sc, c1); err != nil {
-							return
-						}
-					}
-					if len(n.child) == 3 {
-						if c2.typ.cat == arrayT && c2.typ.val.id() == n.typ.val.id() ||
-							isByteArray(c1.typ.TypeOf()) && isString(c2.typ.TypeOf()) {
-							n.gen = appendSlice
-						}
-					}
-				case "cap", "copy", "len":
-					n.typ = sc.getType("int")
-				case "complex":
-					c0, c1 := n.child[1], n.child[2]
-					switch t0, t1 := c0.typ.TypeOf(), c1.typ.TypeOf(); {
-					case isFloat32(t0) && isFloat32(t1):
-						n.typ = sc.getType("complex64")
-					case isFloat64(t0) && isFloat64(t1):
-						n.typ = sc.getType("complex128")
-					case c0.typ.untyped && isNumber(t0) && c1.typ.untyped && isNumber(t1):
-						n.typ = &itype{cat: valueT, rtype: complexType}
-					case c0.typ.untyped && isFloat32(t1) || c1.typ.untyped && isFloat32(t0):
-						n.typ = sc.getType("complex64")
-					case c0.typ.untyped && isFloat64(t1) || c1.typ.untyped && isFloat64(t0):
-						n.typ = sc.getType("complex128")
-					default:
-						err = n.cfgErrorf("invalid types %s and %s", t0.Kind(), t1.Kind())
-					}
-				case "real", "imag":
-					switch k := n.child[1].typ.TypeOf().Kind(); {
-					case k == reflect.Complex64:
-						n.typ = sc.getType("float32")
-					case k == reflect.Complex128:
-						n.typ = sc.getType("float64")
-					case n.child[1].typ.untyped && isNumber(n.child[1].typ.TypeOf()):
-						n.typ = &itype{cat: valueT, rtype: floatType}
-					default:
-						err = n.cfgErrorf("invalid complex type %s", k)
-					}
-				case "make":
-					if n.typ = sc.getType(n.child[1].ident); n.typ == nil {
-						if n.typ, err = nodeType(interp, sc, n.child[1]); err != nil {
-							return
-						}
-					}
-					n.child[1].val = n.typ
-					n.child[1].kind = basicLit
-				case "new":
-					if n.typ, err = nodeType(interp, sc, n.child[1]); err != nil {
-						return
-					}
-					n.typ = &itype{cat: ptrT, val: n.typ}
-				case "recover":
-					n.typ = sc.getType("interface{}")
+				if n.typ, err = nodeType(interp, sc, n); err != nil {
+					return
 				}
-				if n.typ != nil {
-					n.findex = sc.add(n.typ)
-				} else {
+				if n.typ.cat == builtinT {
 					n.findex = -1
 					n.val = nil
+				} else {
+					n.findex = sc.add(n.typ)
 				}
 			case n.child[0].isType(sc):
 				// Type conversion expression
@@ -1563,10 +1507,6 @@ func isMethod(n *node) bool {
 
 func isMapEntry(n *node) bool {
 	return n.action == aGetIndex && n.child[0].typ.cat == mapT
-}
-
-func isBuiltinCall(n *node) bool {
-	return n.kind == callExpr && n.child[0].sym != nil && n.child[0].sym.kind == bltnSym
 }
 
 func isBinCall(n *node) bool {

--- a/interp/run.go
+++ b/interp/run.go
@@ -1717,6 +1717,11 @@ func appendSlice(n *node) {
 }
 
 func _append(n *node) {
+	if c1, c2 := n.child[1], n.child[2]; len(n.child) == 3 && c2.typ.cat == arrayT && c2.typ.val.id() == n.typ.val.id() ||
+		isByteArray(c1.typ.TypeOf()) && isString(c2.typ.TypeOf()) {
+		appendSlice(n)
+		return
+	}
 	dest := genValue(n)
 	value := genValue(n.child[1])
 	next := getExec(n.tnext)

--- a/interp/type.go
+++ b/interp/type.go
@@ -256,17 +256,78 @@ func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
 		}
 
 	case callExpr:
-		if t, err = nodeType(interp, sc, n.child[0]); err != nil {
-			return nil, err
-		}
-		switch t.cat {
-		case valueT:
-			if t.rtype.NumOut() == 1 {
-				t = &itype{cat: valueT, rtype: t.rtype.Out(0)}
+		if interp.isBuiltinCall(n) {
+			// builtin types are special and may depend from their call arguments
+			t.cat = builtinT
+			switch n.child[0].ident {
+			case "complex":
+				var nt0, nt1 *itype
+				if nt0, err = nodeType(interp, sc, n.child[1]); err != nil {
+					return nil, err
+				}
+				if nt1, err = nodeType(interp, sc, n.child[2]); err != nil {
+					return nil, err
+				}
+				if nt0.incomplete || nt1.incomplete {
+					t.incomplete = true
+				} else {
+					switch t0, t1 := nt0.TypeOf(), nt1.TypeOf(); {
+					case isFloat32(t0) && isFloat32(t1):
+						t = sc.getType("complex64")
+					case isFloat64(t0) && isFloat64(t1):
+						t = sc.getType("complex128")
+					case nt0.untyped && isNumber(t0) && nt1.untyped && isNumber(t1):
+						t = &itype{cat: valueT, rtype: complexType}
+					case nt0.untyped && isFloat32(t1) || nt1.untyped && isFloat32(t0):
+						t = sc.getType("complex64")
+					case nt0.untyped && isFloat64(t1) || nt1.untyped && isFloat64(t0):
+						t = sc.getType("complex128")
+					default:
+						err = n.cfgErrorf("invalid types %s and %s", t0.Kind(), t1.Kind())
+					}
+				}
+			case "real", "imag":
+				if t, err = nodeType(interp, sc, n.child[1]); err != nil {
+					return nil, err
+				}
+				if !t.incomplete {
+					switch k := t.TypeOf().Kind(); {
+					case k == reflect.Complex64:
+						t = sc.getType("float32")
+					case k == reflect.Complex128:
+						t = sc.getType("float64")
+					case t.untyped && isNumber(t.TypeOf()):
+						t = &itype{cat: valueT, rtype: floatType}
+					default:
+						err = n.cfgErrorf("invalid complex type %s", k)
+					}
+				}
+			case "cap", "copy", "len":
+				t = sc.getType("int")
+			case "append", "make":
+				t, err = nodeType(interp, sc, n.child[1])
+			case "new":
+				t, err = nodeType(interp, sc, n.child[1])
+				t = &itype{cat: ptrT, val: t, incomplete: t.incomplete}
+			case "recover":
+				t = sc.getType("interface{}")
 			}
-		default:
-			if len(t.ret) == 1 {
-				t = t.ret[0]
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			if t, err = nodeType(interp, sc, n.child[0]); err != nil {
+				return nil, err
+			}
+			switch t.cat {
+			case valueT:
+				if t.rtype.NumOut() == 1 {
+					t = &itype{cat: valueT, rtype: t.rtype.Out(0)}
+				}
+			default:
+				if len(t.ret) == 1 {
+					t = t.ret[0]
+				}
 			}
 		}
 
@@ -471,6 +532,14 @@ func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
 	}
 
 	return t, err
+}
+
+func (interp *Interpreter) isBuiltinCall(n *node) bool {
+	if n.kind != callExpr {
+		return false
+	}
+	s := interp.universe.sym[n.child[0].ident]
+	return s != nil && s.kind == bltnSym
 }
 
 // struct name returns the name of a struct type

